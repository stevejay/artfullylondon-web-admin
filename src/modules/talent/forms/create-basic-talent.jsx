import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Field, reduxForm, formValueSelector } from 'redux-form'

import Form from '_src/shared/components/form'
import FormRow from '_src/shared/components/form/row'
import TextField from '_src/shared/components/text/field'
import FormButtons from '_src/shared/components/form/buttons'
import Divider from '_src/shared/components/divider'
import SelectField from '_src/shared/components/select/field'
import * as talentConstants from '../constants'
import talentType from '_src/domain/types/talent-type'

export class CreateBasicTalentForm extends React.PureComponent {
  handleChangeTalentType = (_, newValue) => {
    if (newValue === talentType.GROUP) {
      this.props.change('firstNames', '')
    }
  }
  render () {
    const {
      submitting,
      talentTypeValue,
      handleSubmit,
      onCancel,
      className
    } = this.props

    const isGroup = talentTypeValue === talentType.GROUP

    return (
      <Form className={className} onSubmit={handleSubmit}>
        <Divider />
        <FormRow>
          <Field
            label='Talent Type'
            name='talentType'
            component={SelectField}
            options={talentConstants.TALENT_TYPES_DROPDOWN_OPTIONS}
            required
            searchable={false}
            onChange={this.handleChangeTalentType}
          />
        </FormRow>
        <FormRow>
          <Field
            label='First Names'
            name='firstNames'
            component={TextField}
            disabled={isGroup}
            forceSingleLine
            autos={false}
            maxLength={
              talentConstants.TALENT_CONSTRAINT.firstNames.length.maximum
            }
          />
          <Field
            label='Last Name / Group Name'
            name='lastName'
            component={TextField}
            required
            forceSingleLine
            autos={false}
            maxLength={
              talentConstants.TALENT_CONSTRAINT.lastName.length.maximum
            }
          />
        </FormRow>
        <FormRow>
          <Field
            label='Common Role'
            name='commonRole'
            component={TextField}
            required
            forceSingleLine
            maxLength={
              talentConstants.TALENT_CONSTRAINT.commonRole.length.maximum
            }
          />
        </FormRow>
        <Divider />
        <FormButtons
          submitLabel='Submit'
          submitting={submitting}
          onCancel={onCancel}
        />
      </Form>
    )
  }
}

CreateBasicTalentForm.propTypes = {
  initialValues: PropTypes.object.isRequired,
  className: PropTypes.string,
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  talentTypeValue: PropTypes.string,
  submitting: PropTypes.bool.isRequired,
  error: PropTypes.any,
  handleSubmit: PropTypes.func.isRequired,
  change: PropTypes.func.isRequired
}

const WrappedEditTalentForm = reduxForm({
  form: talentConstants.EDIT_TALENT_FORM_NAME,
  enableReinitialize: true
})(CreateBasicTalentForm)

const selector = formValueSelector(talentConstants.EDIT_TALENT_FORM_NAME)

export default connect(
  /* istanbul ignore next */
  state => ({ talentTypeValue: selector(state, 'talentType') })
)(WrappedEditTalentForm)
