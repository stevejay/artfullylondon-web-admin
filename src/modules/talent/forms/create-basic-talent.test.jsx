import React from 'react'
import _ from 'lodash'
import { Field } from 'redux-form'

import { CreateBasicTalentForm } from './create-basic-talent'
import talentType from '_src/domain/types/talent-type'

it('should render correctly', () => {
  const wrapper = shallow(
    <CreateBasicTalentForm
      initialValues={{}}
      onSubmit={_.noop}
      onCancel={_.noop}
      talentTypeValue={talentType.INDIVIDUAL}
      submitting={false}
      handleSubmit={_.noop}
      change={_.noop}
    />
  )

  expect(wrapper).toMatchSnapshot()
})

it('should handle a talent type change from individual to group', () => {
  const change = jest.fn()

  const wrapper = shallow(
    <CreateBasicTalentForm
      initialValues={{}}
      onSubmit={_.noop}
      onCancel={_.noop}
      talentTypeValue={talentType.INDIVIDUAL}
      submitting={false}
      handleSubmit={_.noop}
      change={change}
    />
  )

  wrapper.find(Field).at(0).simulate('change', null, talentType.GROUP)

  expect(change).toHaveBeenCalledWith('firstNames', '')
})

it('should handle a talent type change from group to individual', () => {
  const change = jest.fn()

  const wrapper = shallow(
    <CreateBasicTalentForm
      initialValues={{}}
      onSubmit={_.noop}
      onCancel={_.noop}
      talentTypeValue={talentType.GROUP}
      submitting={false}
      handleSubmit={_.noop}
      change={change}
    />
  )

  wrapper.find(Field).at(0).simulate('change', null, talentType.INDIVIDUAL)

  expect(change).not.toHaveBeenCalled()
})
