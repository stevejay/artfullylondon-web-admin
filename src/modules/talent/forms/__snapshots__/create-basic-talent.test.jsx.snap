// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Form
  onSubmit={[Function]}
>
  <Divider
    type="basic"
  />
  <FormRow>
    <Field
      component={[Function]}
      label="Talent Type"
      name="talentType"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "Individual",
            "value": "Individual",
          },
          Object {
            "label": "Group",
            "value": "Group",
          },
        ]
      }
      required={true}
      searchable={false}
    />
  </FormRow>
  <FormRow>
    <Field
      autos={false}
      component={[Function]}
      disabled={false}
      forceSingleLine={true}
      label="First Names"
      maxLength={200}
      name="firstNames"
    />
    <Field
      autos={false}
      component={[Function]}
      forceSingleLine={true}
      label="Last Name / Group Name"
      maxLength={200}
      name="lastName"
      required={true}
    />
  </FormRow>
  <FormRow>
    <Field
      component={[Function]}
      forceSingleLine={true}
      label="Common Role"
      maxLength={200}
      name="commonRole"
      required={true}
    />
  </FormRow>
  <Divider
    type="basic"
  />
  <FormButtons
    onCancel={[Function]}
    submitLabel="Submit"
    submitting={false}
  />
</Form>
`;
